File: problems-originals/0201-0300/0225_implement-stack-using-queues/solution.js
// One queue, rotated on push: the front is always the stack top, so
// pop/top/empty are single queue operations on the front.
class MyStack {
    constructor() {
        this.queue = [];
    }

    push(x) {
        this.queue.push(x);
        // Requeue everything that was below x, so x reaches the front.
        const rotations = this.queue.length - 1;
        for (let i = 0; i < rotations; ++i) this.queue.push(this.queue.shift());
    }

    pop() {
        return this.queue.shift();
    }

    top() {
        return this.queue[0];
    }

    empty() {
        return this.queue.length === 0;
    }
}

File: problems-originals/0201-0300/0225_implement-stack-using-queues/solution.go
// One queue, rotated on push: the front is always the stack top, so
// pop/top/empty are single queue operations on the front. The queue is a
// slice plus a head index — Go ships no queue type, and the statement allows
// simulating one with a list using only its standard operations.
type MyStack struct {
	queue []int
	head  int
}

func NewMyStackTyped() *MyStack {
	return &MyStack{}
}

func (design *MyStack) push(x int) {
	design.queue = append(design.queue, x)
	// Requeue everything that was below x, so x reaches the front.
	rotations := len(design.queue) - design.head - 1
	for i := 0; i < rotations; i++ {
		design.queue = append(design.queue, design.queue[design.head])
		design.head++
	}
}

func (design *MyStack) pop() int {
	value := design.queue[design.head]
	design.head++
	// Once fully drained the consumed prefix is dead; release it.
	if design.head == len(design.queue) {
		design.queue = design.queue[:0]
		design.head = 0
	}
	return value
}

func (design *MyStack) top() int {
	return design.queue[design.head]
}

func (design *MyStack) empty() bool {
	return design.head == len(design.queue)
}
